### PR TITLE
Fix missing interpolation on next/head exports

### LINKED ISSUE #36183: [next/head] [ESM] Element type is invalid: expected a string
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```console
    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Home
    Binaries:
      Node: 16.14.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.1.6-canary.1
      react: 17.0.2
      react-dom: 17.0.2
```

### What browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

### Describe the Bug

Error when rendering _next/head_ with ESM enabled:

```console
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
event - compiled client and server successfully in 189 ms (112 modules)
wait  - compiling / (client and server)...
event - compiled client and server successfully in 48 ms (124 modules)
Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.

Check your code at index.js:6.
    at Home
    at App (webpack-internal:///./node_modules/next/dist/pages/_app.js:69:1)
    at StyleRegistry (C:\head-elements-app\node_modules\styled-jsx\dist\index\index.js:671:34)
    at FlushEffectContainer (C:\head-elements-app\node_modules\next\dist\server\render.js:414:37)
    at AppContainer (C:\head-elements-app\node_modules\next\dist\server\render.js:429:29)
    at AppContainerWithIsomorphicFiberStructure (C:\head-elements-app\node_modules\next\dist\server\render.js:460:57)
    at div
    at Body (C:\head-elements-app\node_modules\next\dist\server\render.js:724:21)
wait  - compiling /_error (client and server)...
event - compiled client and server successfully in 26 ms (125 modules)
error - Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.
```

No error if I remove Head tag.

### Expected Behavior

No error

### To Reproduce

1. from [Head Elements](https://github.com/vercel/next.js/tree/canary/examples/head-elements) example:
`yarn create next-app --example head-elements head-elements-app`
2. add `"type": "module"` to `package.json`
3. `yarn dev`
4. go to `localhost:3000`

### PR DESCRIPTION
**Pull request by Brooooooklyn**
fixes https://github.com/vercel/next.js/issues/36183

## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

File: test/e2e/type-module-interop/index.test.ts [hunk 1]
```diff
@@ -12,10 +12,24 @@ describe('Type module interop', () => {
       files: {
         'pages/index.js': `
           import Link from 'next/link'
+          import Head from 'next/head'
```
Comment (ijjk) [inline test/e2e/type-module-interop/index.test.ts:15]: It seems this test suite is passing without the changes in this PR, are we missing an assertion to capture the previous failure?

File: test/e2e/type-module-interop/index.test.ts [hunk 2]
```diff
@@ -12,10 +12,24 @@ describe('Type module interop', () => {
...
+          import Script from 'next/script'
 
           export default function Page() { 
             return (
               <>
+                <Head>
+                  <title>This page has a title 🤔</title>
+                  <meta charSet="utf-8" />
+                  <meta name="viewport" content="initial-scale=1.0, width=device-width" />
+                  <Script
```
Comment (ijjk) [inline test/e2e/type-module-interop/index.test.ts:25]: `next/script` doesn't need to be nested inside of `next/head` so we can move it out ([related doc](https://nextjs.org/docs/basic-features/script))

Comment (ijjk) [inline test/e2e/type-module-interop/index.test.ts:25]: `next/script` doesn't need to be nested inside of `next/head` so we can move it out ([related doc](https://nextjs.org/docs/basic-features/script))

File: test/e2e/type-module-interop/index.test.ts (hunk 1 above)
Comment (Brooooooklyn) [reply test/e2e/type-module-interop/index.test.ts:15]: Weird, it failed if I comment out my patch on my machine.

image

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| buildDuration | 15.7s | 15.7s | -43ms |
| buildDurationCached | 6.1s | 6.2s | ⚠️ +123ms |
| nodeModulesSize | 481 MB | 481 MB | ⚠️ +58 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.203 | 3.231 | ⚠️ +0.03 |
| / avg req/sec | 780.48 | 773.75 | ⚠️ -6.73 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.372 | 1.368 | 0 |
| /error-in-render avg req/sec | 1822.04 | 1827.98 | +5.94 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 378 B | ⚠️ +27 B |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.73 kB | 5.73 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.63 kB | 2.63 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ⚠️ +27 B |

</details>

<details>
<summary><strong>Client Build Manifests</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 460 B | -1 B |
| Overall change | 461 B | 460 B | -1 B |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>_buildManifest.js</strong></summary>

```diff
@@ -10,7 +10,7 @@ self.__BUILD_MANIFEST = {
   "/dynamic": [
     "static\u002Fchunks\u002Fpages\u002Fdynamic-2fc05fc466ebee90.js"
   ],
-  "/head": ["static\u002Fchunks\u002Fpages\u002Fhead-96a5d6ed07cf5a83.js"],
+  "/head": ["static\u002Fchunks\u002Fpages\u002Fhead-b3f5cd04a41e9ce4.js"],
   "/hooks": ["static\u002Fchunks\u002Fpages\u002Fhooks-9dfe734f583d4926.js"],
   "/image": ["static\u002Fchunks\u002Fpages\u002Fimage-4ebd299a1253d245.js"],
   "/link": ["static\u002Fchunks\u002Fpages\u002Flink-5b526b867abc83fc.js"],
```
</details>
<details>
<summary>Diff for <strong>head-HASH.js</strong></summary>

```diff
@@ -37,6 +37,9 @@
       /* harmony import */ var next_head__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(
         9008
       );
+      /* harmony import */ var next_head__WEBPACK_IMPORTED_MODULE_1___default = /*#__PURE__*/ __webpack_require__.n(
+        next_head__WEBPACK_IMPORTED_MODULE_1__
+      );
 
       var Page = function() {
         return /*#__PURE__*/ (0,
@@ -46,7 +49,7 @@
             children: [
               /*#__PURE__*/ (0,
               react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx)(
-                next_head__WEBPACK_IMPORTED_MODULE_1__["default"],
+                next_head__WEBPACK_IMPORTED_MODULE_1___default(),
                 {
                   children: /*#__PURE__*/ (0,
                   react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx)("title", {
@@ -73,7 +76,9 @@
       __unused_webpack_exports,
       __webpack_require__
     ) {
-      module.exports = __webpack_require__(3121);
+      var head = __webpack_require__(3121);
+      Object.assign(head.default, head);
+      module.exports = head.default;
 
       /***/
     }
```
</details>
</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| buildDuration | 18s | 18s | -52ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +27ms |
| nodeModulesSize | 481 MB | 481 MB | ⚠️ +58 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.232 | 3.234 | 0 |
| / avg req/sec | 773.44 | 773.07 | ⚠️ -0.37 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.397 | 1.371 | -0.03 |
| /error-in-render avg req/sec | 1789.2 | 1822.94 | +33.74 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 385 B | ⚠️ +34 B |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.78 kB | 5.78 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.74 kB | 2.74 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.4 kB | 16.4 kB | ⚠️ +34 B |

</details>

<details>
<summary><strong>Client Build Manifests</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| _buildManifest.js gzip | 457 B | 456 B | -1 B |
| Overall change | 457 B | 456 B | -1 B |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix-head-interpolate | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 542 B | 542 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>_buildManifest.js</strong></summary>

```diff
@@ -10,7 +10,7 @@ self.__BUILD_MANIFEST = {
   "/dynamic": [
     "static\u002Fchunks\u002Fpages\u002Fdynamic-2fc05fc466ebee90.js"
   ],
-  "/head": ["static\u002Fchunks\u002Fpages\u002Fhead-96a5d6ed07cf5a83.js"],
+  "/head": ["static\u002Fchunks\u002Fpages\u002Fhead-b3f5cd04a41e9ce4.js"],
   "/hooks": ["static\u002Fchunks\u002Fpages\u002Fhooks-9dfe734f583d4926.js"],
   "/image": ["static\u002Fchunks\u002Fpages\u002Fimage-4ebd299a1253d245.js"],
   "/link": ["static\u002Fchunks\u002Fpages\u002Flink-5b526b867abc83fc.js"],
```
</details>
<details>
<summary>Diff for <strong>head-HASH.js</strong></summary>

```diff
@@ -37,6 +37,9 @@
       /* harmony import */ var next_head__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(
         9008
       );
+      /* harmony import */ var next_head__WEBPACK_IMPORTED_MODULE_1___default = /*#__PURE__*/ __webpack_require__.n(
+        next_head__WEBPACK_IMPORTED_MODULE_1__
+      );
 
       var Page = function() {
         return /*#__PURE__*/ (0,
@@ -46,7 +49,7 @@
             children: [
               /*#__PURE__*/ (0,
               react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx)(
-                next_head__WEBPACK_IMPORTED_MODULE_1__["default"],
+                next_head__WEBPACK_IMPORTED_MODULE_1___default(),
                 {
                   children: /*#__PURE__*/ (0,
                   react_jsx_runtime__WEBPACK_IMPORTED_MODULE_0__.jsx)("title", {
@@ -73,7 +76,9 @@
       __unused_webpack_exports,
       __webpack_require__
     ) {
-      module.exports = __webpack_require__(3121);
+      var head = __webpack_require__(3121);
+      Object.assign(head.default, head);
+      module.exports = head.default;
 
       /***/
     }
```
</details>
</details>
Commit: fee601b31616e366682d54970267feba3e7a6b05

File: test/e2e/type-module-interop/index.test.ts (hunk 1 above)
Comment (ijjk) [reply test/e2e/type-module-interop/index.test.ts:15]: Oh was testing with the start test mode, does seem to fail without in dev mode 👍

**Review by ijjk — APPROVED** [review]

**Comment by SukkaW** [comment]
It seems that `packages/next/dynamic.js` has the same issue.

**Pull request merged**